File: android/app/src/main/java/com/faceverify/app/credential/Base45.kt
package com.faceverify.app.credential

/** Base45 (RFC 9285) decoder - the QR-alphanumeric-friendly encoding used by
 *  FV1 credentials (same family as EU DCC). Strict: any character outside the
 *  alphabet or an invalid final chunk throws [IllegalArgumentException]. */
object Base45 {
    private const val ALPHABET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ \$%*+-./:"
    private val REVERSE = IntArray(128) { -1 }.also {
        ALPHABET.forEachIndexed { i, c -> it[c.code] = i }
    }

    fun decode(text: String): ByteArray {
        require(text.length % 3 != 1) { "invalid base45 length" }
        val vals = IntArray(text.length) {
            val c = text[it].code
            val v = if (c < 128) REVERSE[c] else -1
            require(v >= 0) { "invalid base45 character '${text[it]}'" }
            v
        }
        val out = ByteArray(text.length / 3 * 2 + if (text.length % 3 == 2) 1 else 0)
        var o = 0
        var i = 0
        while (i + 2 < vals.size) {
            val n = vals[i] + vals[i + 1] * 45 + vals[i + 2] * 45 * 45
            require(n <= 0xFFFF) { "invalid base45 triple" }
            out[o++] = (n ushr 8).toByte()
            out[o++] = n.toByte()
            i += 3
        }
        if (vals.size % 3 == 2) {
            val n = vals[vals.size - 2] + vals[vals.size - 1] * 45
            require(n <= 0xFF) { "invalid base45 pair" }
            out[o] = n.toByte()
        }
        return out
    }
}

File: biometric/core/base45.py
"""Base45 codec (RFC 9285) - the alphanumeric-QR-friendly encoding used by
EU DCC-style credentials. QR alphanumeric mode packs these 45 characters at
5.5 bits each, so base45 payloads fit ~30% more data per QR than base64.

Self-contained (no dependency); strict decode - any character outside the
alphabet or an invalid final chunk raises ``Base45Error``.
"""

from __future__ import annotations

_ALPHABET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ $%*+-./:"
_REVERSE = {c: i for i, c in enumerate(_ALPHABET)}


class Base45Error(ValueError):
    """Input is not valid base45."""


def encode(data: bytes) -> str:
    out = []
    for i in range(0, len(data) - 1, 2):
        n = data[i] * 256 + data[i + 1]
        c, n = divmod(n, 45 * 45)
        b, a = divmod(n, 45)
        out.append(_ALPHABET[a] + _ALPHABET[b] + _ALPHABET[c])
    if len(data) % 2:
        b, a = divmod(data[-1], 45)
        out.append(_ALPHABET[a] + _ALPHABET[b])
    return "".join(out)


def decode(text: str) -> bytes:
    if len(text) % 3 == 1:
        raise Base45Error("invalid base45 length")
    try:
        vals = [_REVERSE[c] for c in text]
    except KeyError as exc:
        raise Base45Error(f"invalid base45 character: {exc.args[0]!r}") from exc
    out = bytearray()
    for i in range(0, len(vals) - 2, 3):
        n = vals[i] + vals[i + 1] * 45 + vals[i + 2] * 45 * 45
        if n > 0xFFFF:
            raise Base45Error("invalid base45 triple")
        out += n.to_bytes(2, "big")
    if len(vals) % 3 == 2:
        n = vals[-2] + vals[-1] * 45
        if n > 0xFF:
            raise Base45Error("invalid base45 pair")
        out.append(n)
    return bytes(out)
